### ISSUE TITLE
UNIX porting

### ISSUE THREAD

**Issue by TheFausap** [JonathanBelanger/DECaxp#4]
Hello Jonathan,

I did some small changes in order to compile the code under LINUX (Opensuse).
I created a fork of your repo, and I committed the changes inside.

Everything compile fine, but some tests fail.

1) AXP_21264_Cache_Test.exe : fails with segmentation fault (just print the banner)

2) AXP_21264_Dump_Test.exe : works

3) AXP_21264_FloatingPointTest.exe : 2 tests fails (nan related), this is the output:
```
AXP 21264 IEEE Floating Point Tester
    234: isSigned: -nan, expecting 1 --> failed (0xfff8000020000000)
    255: isSigned: -nan, expecting 1 --> failed (0xfff8000020000000)

3904 tests passed, 2 failed, and 22260 skipped, with a total of 26166 tests.
```
I'm worried about all the skipped tests, is it normal ? 

4) AXP_21264_Icache_Test.exe : works (I suppose). Hits percentage is 99.41 %

5) AXP_21264_IntegerLoadTest.exe : works. All the test cases passed.

6) AXP_21264_Prediction_Test.exe : works

7) AXP_Disk_Test.exe : fails (because there's an hardcoded path). I'll fix this.

8) AXP_DS12887A_Test.exe : works

9) AXP_Test_Structure_Sizes.exe : fail one check. This is the output:```

AXP_VHDX_META_PAR_HDR    =   20 (  24): failed
        AXP_VHDX_META_PAR_HDR is supposed to be 20 bytes, but the GCC compiler pads this out to 24 bytes (byte boundary)
                locType (len = 16) at offset 0
                res_1 (len = 2) at offset 16
                keyValCnt (len = 2) at offset 18
```
Maybe the problem is the GCC compiler too much different from the one you used (6.x)
```
10) AXP_Test_WinPcap.exe : works (I suppose). It lists all the network adapters present. This is a partial output:
```
1: wlo1
        Description: No description available
        Loopback: No
        Addresses:
                Address Family: #17
                        Address Family Name: Unknown
                Address Family: #2
                        Address Family Name: AF_INET
                        Address: 172.20.10.2
                        Netmask: 255.255.255.240
                        Broadcast Address: 172.20.10.15
                Address Family: #10
                        Address Family Name: AF_INET6
                        Address: fe80::7e3e:39ff:5609:d5ce%wlo1
```
wlo1 is the wireless adapter on my laptop and also loopback is present in the list.

**Comment by TheFausap** [comment]
About the AXP_Disk_Test.exe, it seems there's a deeper problem opening the file.
I added some DEBUG information, and I have this output, at the moment:
```
Test 13: Printable from 'y' to '~' and Nonprintable characters: [actual=d579c862, expected=d579c862] - Passed
Test 14: Nonprintable characters from 0xa1 to 0xc8: [actual=ba979ad0, expected=ba979ad0] - Passed
Test 15: Nonprintable characters from 0xc9 to 0xf0: [actual=2b29d913, expected=2b29d913] - Passed
Test 16: Nonprintable and all the printable characters: [actual=24c5d375, expected=24c5d375] - Passed
Test 17: The string '123456789': [actual=e3069283, expected=e3069283] - Passed
Test 18: Create a VHDX(v1) disk in /tmp with the name of RZ1CD-CS.vhdx of 3145728 bytes in size...
Checking the UUID with the KNOWN UUID list
Validated Type VHDX
Version 1 requested
Validate create finished
Calling the backend VHDX function
VOID Block allocation
VHDX file name : /tmp/RZ1CD-CS.vhdx
VHDX Block allocation
VOID Block allocation
Segmentation fault (core dumped)
```

**Comment by TheFausap** [comment]
I found the place where the segfault occurs.
In AXP_Disk_Test.c, the function AXP_Allocate_Block (during the VHDX test) is called three times.
The third time is called in this way:
`vhdx->filePath = AXP_Allocate_Block(-(strlen(path) + 1));`
now in this case in gdb I have this:
```
Breakpoint 1, AXP_Allocate_Block (blockType=-19)
    at /home/fausap/DECaxp/src/comutl/AXP_Blocks.c:121
```
now because replaceBlk is not NULL in this case
`2: replaceBlk = (void *) 0x409c11 <AXP_TraceInit_Once>`
a memcpy is called (at line 348 of my version of AXP_Blocks.c) with negative size (so the segfault):

```
memcpy(
			    replaceBlk,
			    allocBlock,
			    (head->size - sizeof(AXP_BLOCK_TL)));
```
head->size is 0.
I do not understand exactly what AXP_Allocate_Block does, so I cannot understand if replaceBlk is ok to be AXP_TraceInit_Once.

**Comment by JonathanBelanger** [comment]
The negative size is used to determine that a non-structured block of memory if being requested.  The call to calloc (and I always use calloc) should take the absolute value.  If it doesn't, then this is a bug.

Right now, I'm working on an Open version of SDL (Structure Definition Language) that should simplify my structure definitions.  As for your tests that failed.  The floating point ones are expected.  There is a bug in the test data set where a test case is executed at two different times and the expected results are different.  The cache one, I'm not surprised.  That is old test code and I'll be changing the caching scheme (one Bcache shared across all CPUs so that there does not need to be a message passing scheme to keep each CPU's Bcache in sync with the others).  The winpcap one is exceptionally limited.  I was going to expand upon it at a future date.

Did you have to make any code changes?  I'm trying to compile with c99 portability, which should limit the number of items that need changing.

Thanks,

~Jon.

**Comment by TheFausap** [comment]
The real changes are in AXP_Ethernet.c, because winpcap uses some nonstandard pcap routines.
This is the patch for that file.

```
--- /home/fausap/DECaxp.orig/src/comutl/AXP_Ethernet.c  2018-08-19 00:11:08.587366340 +0200
+++ AXP_Ethernet.c      2018-09-04 22:34:49.363675208 +0200
@@ -51,6 +51,12 @@
  *  ~NULL:     A pointer to the Ethernet Handle through which packets are
  *             send and received.
  */
+
+#ifndef WINPCAP
+#define PCAP_OPENFLAG_PROMISCUOUS 1
+#define PCAP_OPENFLAG_NOCAPTURE_LOCAL 8
+#endif
+
 AXP_Ethernet_Handle *AXP_EthernetOpen(char *name, u8 cardNo)
 {
     AXP_Ethernet_Handle *retVal;
@@ -58,14 +64,19 @@
     retVal = AXP_Allocate_Block(AXP_ETHERNET_BLK);
     if (retVal != NULL)
     {
-
+#ifdef WINPCAP
        retVal->handle = pcap_open(
+#else
+       retVal->handle = pcap_open_live(
+#endif
                                name,
                                SIXTYFOUR_K,
                                (PCAP_OPENFLAG_PROMISCUOUS |
                                    PCAP_OPENFLAG_NOCAPTURE_LOCAL),
                                AXP_ETH_READ_TIMEOUT,
+#ifdef WINPCAP
                                NULL,
+#endif
                                retVal->errorBuf);
        if (retVal->handle == NULL)
        {

```
and as said by @FreeBSD-pzn, you need to change the include filename in AXP_21264_Ebox_PALfunctions.c, because it's case sensitive.

```
--- /home/fausap/DECaxp.orig/src/cpu/AXP_21264_Ebox_PALfunctions.c      2018-08-19 00:11:08.595366245 +0200
+++ AXP_21264_Ebox_PALfunctions.c       2018-09-04 22:34:49.375675040 +0200
@@ -26,7 +26,7 @@
  */
 
 #include "AXP_Configure.h"
-#include "AXP_21264_Ebox_PALFunctions.h"
+#include "AXP_21264_Ebox_PALfunctions.h"
 
 /*
  * AXP_HWLD

```
I also change a little bit the Makefile, to include libraries needed under LINUX and some define fort the LIBPCAP.

```
PCAP = -DLIBPCAP
CFLAGS = $(INCS) $(PCAP) -m64 -std=gnu99 -Wall -pthread -D_AXP_COMPILER_PRESENT_
LFLAGS = -lxml2 -lm -lrt -lpcap -luuid
```

Other changes I did, they are just for the AXP_Disk debugging.

About the AXP_Allocate_Block, I see you do a calloc with the abs value, indeed this is not a problem.
The memcpy has this issue, because the head->size is 0.
Just to understand the idea behind your code:

In the function _AXP_VHDX_Create (file: AXP_VHDX.c) you call:
  1. outBuf = AXP_Allocate_Block(-SIXTYFOUR_K) at line 156
      and this allocate a block of 64KB of non structured memory as outBuf.
  2. vhdx = (AXP_VHDX_Handle *) AXP_Allocate_Block(AXP_VHDX_BLK) at line 163
      and this allocate a block of structured memory for the VHDX
  3. vhdx->filePath = AXP_Allocate_Block(-(strlen(path) + 1)) at line 166
      this I don't understand.

The call at 3 generates the segfault. Going to the file AXP_Blocks.c, the first two calls are fine. The third one has a big difference with the other VOID block call, that generates a different code path causing the segfault.

   1. blockType = AXP_VOID_BLK, replaceBlk = NULL
   2. blockType = AXP_VHDX_BLK, replaceBlk = NULL
   3. blockType = AXP_VOID_BLK, replaceBlk = <AXP_TraceInit_Once>

the not-NULL replaceBlk calls the wrong memcpy at line 327 and then fault!

But why this call at point 3 should replace an existing block ? Is it intended ?

**Comment by JonathanBelanger** [comment]
I'll look into the AXP_Blocks issue.

As for the conditional compiling, I will do this a different way.  I prefer to put all conditional compilations into header files and leave the source files as close to clean as possible.  If different header files are needed, then those will also go into appropriate header files.

Thanks for the analysis.

~Jon.

**Comment by JonathanBelanger** [comment]
I'm in the process of getting the code to compile with a later version of GCC and now clang-9.  Both have called out a number of issues that previous versions of GCC did not.  Once I get clean compiles for both of these compilers, I will be rerunning and fixing all the tests.  In the meantime, I am going to keep this issue open.  I have also converted to using CMake, which will give more flexibility with target platforms moving forward.

**Comment by TheFausap** [comment]
Cool! I'll continue to do some tests in parallel.

**Comment by JonathanBelanger** [comment]
Hi Fausto,

All the code compiles, even using clang-9, though executing clang built code on Cygwin on Windows generates errors, which are not seen when compiling with gcc.  All test programs work as expected.   Now for some of your previous points:

1. `AXP_21264_Cache_Test.exe` now works.  I had not been maintaining it while working in the cache code.
1. `AXP_21264_Dump_Test.exe`  I broke this and have not fixed it.  It is working as expected.
1. `AXP_21264_FloatingPointTest.exe` is expected to fail 2 of its tests.  If you look at the other executed test cases, both of these are executed prior and pass.  These 2 test cases may or may not be correct.  I contacted the author of the test sets, and he acknowledge the issue, but did not have and was not going to change the data sets.  As for running the other data sets, I have plans to do so (not sure when this will happen - if you want to try, please feel free to do so).
1. `AXP_21264_Icache_Test.exe` the I-cache and D-cache code are based off of the actual implementation within in the Alpha CPU.  I think I can do better from a performance perspective, but don't expected to do much better with the hit rate.
1. `AXP_Disk_Test.exe` I fixed a number of things in the Virtual Disk code.  I had made some changes, as you may be aware, to the AXP_Allocate_Block code.  In the cases where a negative value is passed as the block-type, either a `NULL` or the address of the current block so that it can be freed prior to being reallocated.  There was a number of calls that needed to be corrected throughout the code.  I'd love to have a whole bunch more test cases for this test.  Each VHD type should be tested and read and write cases, as well as performance cases would be ideal.  I did do some looking around for test cases, but could not find anything in the limited time I did my searches.
1. `AXP_Test_Structure_Sizes.exe` has been updated to indicate that a difference in sizeof versus actual layout of the structure.  It now displays an `expected` not and considers the result a passed tests.
1. `AXP_Test_WinPcap.exe` I admit is a pretty lame test.  If you want to add more cases to this, please feel free to do so.  When you have something, submit a pull request.  Also, I was going to add your patch, but when I looked for the `WINPCAP` macro definition in the pcap header files, I could not find it.  So, I did not add this patch.

*Finally*, thanks for picking this up and submitting detailed descriptions of the issues you ran into.  It is very much appreciated.

**Comment by TheFausap** [comment]
Hi Jonathan,

very nice update and you're welcome! I love your passion for DEC, and I totally agree with you!
Anyway, about Winpcap, I think you should move to NPCAP (https://nmap.org/npcap/), the evolution of winpcap (no longer developed and maintained) that support Windows 10 too.
And, but I have to check this, it seems now more "compatible" with pcap library under UNIX, if you look at pcap.h header file (in npcap sdk)

```
#if defined(_WIN32)
  #include <winsock2.h>		/* u_int, u_char etc. */
  #include <io.h>		/* _get_osfhandle() */
#elif defined(MSDOS)
  #include <sys/types.h>	/* u_int, u_char etc. */
  #include <sys/socket.h>
#else /* UN*X */
  #include <sys/types.h>	/* u_int, u_char etc. */
  #include <sys/time.h>
#endif /* _WIN32/MSDOS/UN*X */
```

What do you think?

**Comment by TheFausap** [comment]
Hi Jonathan,

I replicated all my changes in a forked repo, instead of using the WINPCAP define, I preferred using "linux" one, automatically defined by gcc.
Everything is ok, I got only a segmentation fault in the cache test, but I fixed this commenting the line 

`line = AXP_Allocate_Block(-lineLen);`

I didn't investigate the reason yet.

**Comment by JonathanBelanger** [comment]
That line of code should be changed to:
```
line = AXP_Allocate_block(-lineLen, NULL);
```
I missed that one.

**Comment by JonathanBelanger** [comment]
Also, in looking over the code, I realized that the AXP_Test_WinPcap, now called AXP_Test_Npcap, test code and the AXP_Ethernet.c code are not really fully functional.  The test code was just my attempt to learn something about the WinPcap, and now Npcap, APIs.  Since I moved to Npcap, the test code is not really working.  Also, the AXP_Ethernet interface needs to be fully defined, which I am not quite ready for.
